### DISCUSSION TITLE
How to increase pytorch generation speed?

### DISCUSSION THREAD

**Discussion by bendoesai** [Trusted-AI/adversarial-robustness-toolbox#2597] [Q&A] [answered]
I'm running large-scale adversarial sample generation on many images/videos using Pytorch and AutoPGD. However, generation still takes a very long time (1-5+ days on 22388 samples depending on model), despite running on top-of-the-line GPUs (a100, h100, etc) with plenty of memory. Below are my configurations for both the PyTorch Classifier and AutoPGD attack. Perhaps it has something to do with numpy arrays being CPU data and my model being on the GPU? Are these runtimes for a decently large model reasonable?

What are some ways I can improve the generation speed given access to lots of compute? In addition, is there any way to integrate with a pytorch dataloader for just adversarial generation, or do I have to keep collating the dataloaders I'm given into a big numpy array?

```
classifier = PyTorchClassifier(
            model=art_model,
            clip_values=((0-mean)/std, (1-mean)/std),
            loss=art_model.loss_func,
            input_shape=(3, 256, 256),
            nb_classes=2,
            device_type= 'gpu' if torch.cuda.is_available() else 'cpu',
        )

AutoProjectedGradientDescent(
            estimator=classifier,
            eps=0.03,
            eps_step=0.1,
            max_iter=100,
            targeted=False,
            nb_random_init=5,
            batch_size=32,
            loss_type="cross_entropy",
        )
```

**Comment by bendoesai** [comment, marked as answer]
an update: I upgraded to ART 1.19 and everything is much smoother now that I can take advantage of multiprocessing with AutoAttack. Should have considered this sooner. Thanks for maintaining the library!
